dragonfly/actions/keyboard.py

```diff
@@ -34,68 +31,94 @@
+                repr(self.events()))
 
     def on_events(self, timeout=0):
-        events = [(m, True, 0) for m in self._modifiers]
-        events.append((self._code, True, timeout))
+        """Return events for pressing this key down."""
+        if self._is_text:
+            events = [(self._code, True, timeout, True)]
+        else:
+            events = [(m, True, 0) for m in self._modifiers]
+            events.append((self._code, True, timeout))
         return events
 
     def off_events(self, timeout=0):
-        events = [(m, False, 0) for m in self._modifiers]
-        events.append((self._code, False, timeout))
-        events.reverse()
+        """Return events for releasing this key."""
+        if self._is_text:
+            events = [(self._code, False, timeout, True)]
+        else:
+            events = [(m, False, 0) for m in self._modifiers]
+            events.append((self._code, False, timeout))
+            events.reverse()
         return events
 
     def events(self, timeout=0):
-        events = [(self._code, True, 0), (self._code, False, timeout)]
-        for m in self._modifiers[-1::-1]:
-            events.insert(0, (m, True, 0))
-            events.append((m, False , 0))
+        """Return events for pressing and then releasing this key."""
+        if self._is_text:
+            events = [(self._code, True, timeout, True),
+                      (self._code, False, timeout, True)]
+        else:
+            events = [(self._code, True, 0), (self._code, False, timeout)]
+            for m in self._modifiers[-1::-1]:
+                events.insert(0, (m, True, 0))
+                events.append((m, False, 0))
         return events
 
 
-#---------------------------------------------------------------------------
-# Keyboard access class.
-
 class Keyboard(object):
+    """Static class wrapper around SendInput."""
 
-    shift_code =    win32con.VK_SHIFT
-    ctrl_code =     win32con.VK_CONTROL
-    alt_code =      win32con.VK_MENU
+    shift_code = win32con.VK_SHIFT
+    ctrl_code = win32con.VK_CONTROL
+    alt_code = win32con.VK_MENU
 
     @classmethod
     def send_keyboard_events(cls, events):
         """
-            Send a sequence of keyboard events.
+        Send a sequence of keyboard events.
 
-            Positional arguments:
-            events -- a sequence of 3-tuples of the form
-                (keycode, down, timeout), where
-                keycode (int): virtual key code.
+        Positional arguments:
+        events -- a sequence of tuples of the form
+            (keycode, down, timeout), where
+                keycode (int): Win32 Virtual Key code.
                 down (boolean): True means the key will be pressed down,
                     False means the key will be released.
                 timeout (int): number of seconds to sleep after
                     the keyboard event.
-
+            or
+            (character, down, timeout, is_text), where
```

Comment:
This doesn't exactly match how the code behaves: it actually supports lists of length 3 and 4, and will break if the user passes in False for is_text. I think we should change the code to match this documentation (it's confusing to have the multiple length special case, and will break if we ever add another parameter).

Reply:
Most of this code is based on something very hack-ish I did a couple of years ago. Honestly, it should have been rewritten. I'm entirely open to doing just that. Due to the backwards compatibility issues, I think this should be hot fixed to be disabled by default. The rest of it can be fixed in the ways described.